walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,143 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
+
+  /* runs in O(n*m+p), where n is the number of events, m is the maximum number of attendees for any given event, and p is
+   * the number of attendees in the request
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // find all the viable meeting times for the mandatory attendees
+    Collection<TimeRange> mandatoryViableMeetingTimes = getViableMeetingTimes(events, duration, attendees);
+    // find all the viable meeting times for the optional attendees
+    Collection<TimeRange> optionalViableMeetingTimes = getViableMeetingTimes(events, duration, optionalAttendees);
+    // see if there are any overlapping meeting times between what the mandatory and optional attendees can attend
+    Collection<TimeRange> intersectionMeetingTimes = intersectionMeetingTimes(duration, mandatoryViableMeetingTimes, optionalViableMeetingTimes);
+
+    // if there is no meeting time where all of the mandatory and optional attendees can attend, return only the times when the mandatory employees are available
+    if(intersectionMeetingTimes.size() == 0) {
+        // case where there are only optional attendees, and there is not a time where all of the optional attendees can attend
+        if(attendees.size() == 0) {
+            return Arrays.asList();
+        }
+        return mandatoryViableMeetingTimes;
+    }
+    else {
+        return intersectionMeetingTimes;
+    }
+  }
+
+  private static Collection<TimeRange> getViableMeetingTimes(Collection<Event> events, long duration, Collection<String> attendees) {
+    // used in the function intersection
+    HashMap<String, Boolean> attendeesMap = new HashMap<String, Boolean>();
+    Iterator iterator = attendees.iterator(); 
+    while(iterator.hasNext()) {
+        attendeesMap.put((String) iterator.next(), true);
+    }
+
+    // all of the minutes when meetings can be held will be marked true
+    boolean[] viableTimes = new boolean[TimeRange.END_OF_DAY+1];
+
+    // initially, all of the minutes are viable
+    for(int i=0;i<viableTimes.length;i++) {
+        viableTimes[i] = true;
+    }
+
+    // consider each event independently
+    iterator = events.iterator(); 
```

Comment:
Better to create a different iterator variable and name appropriately instead of reassigning to the variable to avoid confusion what you're iterating over. Ex. attendeesIterator and eventsIterator.